File: lib/VM/JSLib/Map.cpp
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

//===----------------------------------------------------------------------===//
/// \file
/// ES6.0 23.1 Initialize the Map constructor.
//===----------------------------------------------------------------------===//

#include "JSLibInternal.h"

#include "hermes/VM/StringPrimitive.h"

namespace hermes {
namespace vm {

HermesValue createMapConstructor(Runtime &runtime) {
  auto mapPrototype = Handle<JSObject>::vmcast(&runtime.mapPrototype);

  struct : public Locals {
    PinnedValue<NativeConstructor> cons;
  } lv;
  LocalsRAII lraii(runtime, &lv);

  // Map.prototype.xxx methods.
  defineMethod(
      runtime,
      mapPrototype,
      Predefined::getSymbolID(Predefined::clear),
      nullptr,
      mapPrototypeClear,
      0);

  defineMethod(
      runtime,
      mapPrototype,
      Predefined::getSymbolID(Predefined::deleteStr),
      nullptr,
      mapPrototypeDelete,
      1);

  runtime.mapPrototypeEntries = defineMethod(
      runtime,
      mapPrototype,
      Predefined::getSymbolID(Predefined::entries),
      nullptr,
      mapPrototypeEntries,
      0);

  defineMethod(
      runtime,
      mapPrototype,
      Predefined::getSymbolID(Predefined::forEach),
      nullptr,
      mapPrototypeForEach,
      1);

  defineMethod(
      runtime,
      mapPrototype,
      Predefined::getSymbolID(Predefined::get),
      nullptr,
      mapPrototypeGet,
      1);

  defineMethod(
      runtime,
      mapPrototype,
      Predefined::getSymbolID(Predefined::has),
      nullptr,
      mapPrototypeHas,
      1);

  defineMethod(
      runtime,
      mapPrototype,
      Predefined::getSymbolID(Predefined::keys),
      nullptr,
      mapPrototypeKeys,
      0);

  runtime.mapPrototypeSet = defineMethod(
      runtime,
      mapPrototype,
      Predefined::getSymbolID(Predefined::set),
      nullptr,
      mapPrototypeSet,
      2);

  defineAccessor(
      runtime,
      mapPrototype,
      Predefined::getSymbolID(Predefined::size),
      nullptr,
      mapPrototypeSizeGetter,
      nullptr,
      false,
      true);

  defineMethod(
      runtime,
      mapPrototype,
      Predefined::getSymbolID(Predefined::values),
      nullptr,
      mapPrototypeValues,
      0);

  DefinePropertyFlags dpf = DefinePropertyFlags::getNewNonEnumerableFlags();
  runtime.ignoreAllocationFailure(JSObject::defineOwnProperty(
      mapPrototype,
      runtime,
      Predefined::getSymbolID(Predefined::SymbolIterator),
      dpf,
      runtime.mapPrototypeEntries));

  dpf = DefinePropertyFlags::getDefaultNewPropertyFlags();
  dpf.writable = 0;
  dpf.enumerable = 0;
  defineProperty(
      runtime,
      mapPrototype,
      Predefined::getSymbolID(Predefined::SymbolToStringTag),
      runtime.getPredefinedStringHandle(Predefined::Map),
      dpf);

  defineSystemConstructor(
      runtime,
      Predefined::getSymbolID(Predefined::Map),
      mapConstructor,
      mapPrototype,
      0,
      lv.cons);

  return lv.cons.getHermesValue();
}

/// Populate the Map with the contents of the source Map.
/// \param target the Map to populate (newly constructed).
/// \param src the Map to pull the entries from.
/// \return the newly populated map.
static ExecutionStatus
mapFromMapFastPath(Runtime &runtime, Handle<JSMap> target, Handle<JSMap> src) {
  // TODO: This can be improved further by avoiding any rehashes and the
  // SmallHermesValue unbox/boxing. We should be able to make an
  // OrderedHashMap::clone that initializes based on an existing Map
  // and clones all entries directly somehow.
  struct : public Locals {
    PinnedValue<> keyHandle;
    PinnedValue<> valueHandle;
  } lv;
  LocalsRAII lraii{runtime, &lv};
  return JSMap::forEachNative(
      src,
      runtime,
      [&target, &lv](
          Runtime &runtime, Handle<HashMapEntry> entry) -> ExecutionStatus {
        lv.keyHandle = entry->key.unboxToHV(runtime);
        lv.valueHandle = entry->value.unboxToHV(runtime);
        if (LLVM_UNLIKELY(
                JSMap::insert(target, runtime, lv.keyHandle, lv.valueHandle) ==
                ExecutionStatus::EXCEPTION)) {
          return ExecutionStatus::EXCEPTION;
        }
        return ExecutionStatus::RETURNED;
      });
}

CallResult<HermesValue> mapConstructor(void *, Runtime &runtime) {
  NativeArgs args = runtime.getCurrentFrame().getNativeArgs();
  GCScope gcScope{runtime};
  if (LLVM_UNLIKELY(!args.isConstructorCall())) {
    return runtime.raiseTypeError("Constructor Map requires 'new'");
  }

  struct : public Locals {
    PinnedValue<JSObject> selfParent;
    PinnedValue<JSMap> self;
    PinnedValue<> adderProp;
    PinnedValue<SymbolID> iteratorSymbol;
    PinnedValue<Callable> iterMethod;
  } lv;
  LocalsRAII lraii(runtime, &lv);
  if (LLVM_LIKELY(
          args.getNewTarget().getRaw() ==
          runtime.mapConstructor.getHermesValue().getRaw())) {
    lv.selfParent = runtime.mapPrototype;
  } else {
    CallResult<PseudoHandle<JSObject>> thisParentRes =
        NativeConstructor::parentForNewThis_RJS(
            runtime,
            Handle<Callable>::vmcast(&args.getNewTarget()),
            runtime.mapPrototype);
    if (LLVM_UNLIKELY(thisParentRes == ExecutionStatus::EXCEPTION)) {
      return ExecutionStatus::EXCEPTION;
    }
    lv.selfParent = std::move(*thisParentRes);
  }

  lv.self = JSMap::create(runtime, lv.selfParent);

  if (LLVM_UNLIKELY(
          JSMap::initializeStorage(lv.self, runtime) ==
          ExecutionStatus::EXCEPTION)) {
    return ExecutionStatus::EXCEPTION;
  }

  if (args.getArgCount() == 0 || args.getArg(0).isUndefined() ||
      args.getArg(0).isNull()) {
    return lv.self.getHermesValue();
  }

  auto propRes = JSObject::getNamed_RJS(
      lv.self, runtime, Predefined::getSymbolID(Predefined::set));
  if (LLVM_UNLIKELY(propRes == ExecutionStatus::EXCEPTION)) {
    return ExecutionStatus::EXCEPTION;
  }

  // ES6.0 23.1.1.1.7: Cache adder across all iterations of the loop.
  lv.adderProp = std::move(*propRes);
  auto adder = Handle<Callable>::dyn_vmcast(Handle<>{lv.adderProp});
  if (!adder) {
    return runtime.raiseTypeError("Property 'set' for Map is not callable");
  }

  lv.iteratorSymbol = Predefined::getSymbolID(Predefined::SymbolIterator);
  auto iterMethodRes =
      getMethod(runtime, args.getArgHandle(0), lv.iteratorSymbol);
  if (LLVM_UNLIKELY(iterMethodRes == ExecutionStatus::EXCEPTION)) {
    return ExecutionStatus::EXCEPTION;
  }
  if (!vmisa<Callable>(iterMethodRes->getHermesValue())) {
    return runtime.raiseTypeError("iterator method is not callable");
  }
  lv.iterMethod.castAndSetHermesValue<Callable>(
      iterMethodRes->getHermesValue());

  // Check and run fast path.
  if (LLVM_LIKELY(
          adder.getHermesValue().getRaw() ==
          runtime.mapPrototypeSet.getHermesValue().getRaw())) {
    // If the iterable is a Map with the original iterator,
    // then we can do for-loop.
    if (Handle<JSMap> inputMap = args.dyncastArg<JSMap>(0); inputMap &&
        LLVM_LIKELY(lv.iterMethod.getHermesValue().getRaw() ==
                    runtime.mapPrototypeEntries.getHermesValue().getRaw())) {
      if (LLVM_UNLIKELY(
              mapFromMapFastPath(runtime, lv.self, inputMap) ==
              ExecutionStatus::EXCEPTION)) {
        return ExecutionStatus::EXCEPTION;
      }
      return lv.self.getHermesValue();
    }

    // TODO: Fast path for JSArray input.
  }

  return addEntriesFromIterable(
      runtime,
      lv.self,
      args.getArgHandle(0),
      Handle<Callable>{lv.iterMethod},
      [&runtime, &self = lv.self, adder](
          Runtime &, Handle<> key, Handle<> value) {
        return Callable::executeCall2(
                   adder,
                   runtime,
                   self,
                   key.getHermesValue(),
                   value.getHermesValue())
            .getStatus();
      });
}

CallResult<HermesValue> mapPrototypeClear(void *, Runtime &runtime) {
  NativeArgs args = runtime.getCurrentFrame().getNativeArgs();
  auto selfHandle = args.dyncastThis<JSMap>();
  if (LLVM_UNLIKELY(!selfHandle)) {
    return runtime.raiseTypeError(
        "Non-Map object called on Map.prototype.clear");
  }
  JSMap::clear(selfHandle, runtime);
  return HermesValue::encodeUndefinedValue();
}

CallResult<HermesValue> mapPrototypeDelete(void *, Runtime &runtime) {
  NativeArgs args = runtime.getCurrentFrame().getNativeArgs();
  auto selfHandle = args.dyncastThis<JSMap>();
  if (LLVM_UNLIKELY(!selfHandle)) {
    return runtime.raiseTypeError(
        "Non-Map object called on Map.prototype.delete");
  }
  return HermesValue::encodeBoolValue(
      JSMap::erase(selfHandle, runtime, args.getArgHandle(0)));
}

CallResult<HermesValue> mapPrototypeEntries(void *, Runtime &runtime) {
  NativeArgs args = runtime.getCurrentFrame().getNativeArgs();
  auto selfHandle = args.dyncastThis<JSMap>();
  if (LLVM_UNLIKELY(!selfHandle)) {
    return runtime.raiseTypeError(
        "Non-Map object called on Map.prototype.entries");
  }
  struct : public Locals {
    PinnedValue<JSMapIterator> iterator;
  } lv;
  LocalsRAII lraii(runtime, &lv);
  lv.iterator.castAndSetHermesValue<JSMapIterator>(
      JSMapIterator::create(
          runtime, Handle<JSObject>::vmcast(&runtime.mapIteratorPrototype))
          .getHermesValue());
  lv.iterator->initializeIterator(runtime, selfHandle, IterationKind::Entry);
  return lv.iterator.getHermesValue();
}

CallResult<HermesValue> mapPrototypeForEach(void *, Runtime &runtime) {
  NativeArgs args = runtime.getCurrentFrame().getNativeArgs();
  auto selfHandle = args.dyncastThis<JSMap>();
  if (LLVM_UNLIKELY(!selfHandle)) {
    return runtime.raiseTypeError(
        "Non-Map object called on Map.prototype.forEach");
  }
  auto callbackfn = args.dyncastArg<Callable>(0);
  if (LLVM_UNLIKELY(!callbackfn)) {
    return runtime.raiseTypeError(
        "callbackfn must be Callable in Map.prototype.forEach");
  }
  auto thisArg = args.getArgHandle(1);
  if (JSMap::forEach(selfHandle, runtime, callbackfn, thisArg) ==
      ExecutionStatus::EXCEPTION)
    return ExecutionStatus::EXCEPTION;
  return HermesValue::encodeUndefinedValue();
}

CallResult<HermesValue> mapPrototypeGet(void *, Runtime &runtime) {
  NativeArgs args = runtime.getCurrentFrame().getNativeArgs();
  auto *self = dyn_vmcast<JSMap>(args.getThisArg());
  if (LLVM_UNLIKELY(!self)) {
    return runtime.raiseTypeError("Non-Map object called on Map.prototype.get");
  }
  return self->get(runtime, args.getArg(0)).unboxToHV(runtime);
}

CallResult<HermesValue> mapPrototypeHas(void *, Runtime &runtime) {
  NativeArgs args = runtime.getCurrentFrame().getNativeArgs();
  auto *self = dyn_vmcast<JSMap>(args.getThisArg());
  if (LLVM_UNLIKELY(!self)) {
    return runtime.raiseTypeError("Non-Map object called on Map.prototype.has");
  }
  return HermesValue::encodeBoolValue(self->has(runtime, args.getArg(0)));
}

CallResult<HermesValue> mapPrototypeKeys(void *, Runtime &runtime) {
  NativeArgs args = runtime.getCurrentFrame().getNativeArgs();
  auto selfHandle = args.dyncastThis<JSMap>();
  if (LLVM_UNLIKELY(!selfHandle)) {
    return runtime.raiseTypeError(
        "Non-Map object called on Map.prototype.keys");
  }

  struct : public Locals {
    PinnedValue<JSMapIterator> iterator;
  } lv;
  LocalsRAII lraii(runtime, &lv);
  lv.iterator.castAndSetHermesValue<JSMapIterator>(
      JSMapIterator::create(
          runtime, Handle<JSObject>::vmcast(&runtime.mapIteratorPrototype))
          .getHermesValue());
  lv.iterator->initializeIterator(runtime, selfHandle, IterationKind::Key);
  return lv.iterator.getHermesValue();
}

// ES12 23.1.3.9 Map.prototype.set ( key, value )
CallResult<HermesValue> mapPrototypeSet(void *, Runtime &runtime) {
  NativeArgs args = runtime.getCurrentFrame().getNativeArgs();
  auto selfHandle = args.dyncastThis<JSMap>();
  if (LLVM_UNLIKELY(!selfHandle)) {
    return runtime.raiseTypeError("Non-Map object called on Map.prototype.set");
  }
  auto keyHandle = args.getArgHandle(0);
  // 5. If key is -0, set key to +0.
  // N.B. in the case of Map, only key should be normalized but not the value.
  auto key = keyHandle->isNumber() && keyHandle->getNumber() == 0
      ? HandleRootOwner::getZeroValue()
      : keyHandle;
  if (LLVM_UNLIKELY(
          JSMap::insert(selfHandle, runtime, key, args.getArgHandle(1)) ==
          ExecutionStatus::EXCEPTION)) {
    return ExecutionStatus::EXCEPTION;
  }
  return selfHandle.getHermesValue();
}

CallResult<HermesValue> mapPrototypeSizeGetter(void *, Runtime &runtime) {
  NativeArgs args = runtime.getCurrentFrame().getNativeArgs();
  auto self = dyn_vmcast<JSMap>(args.getThisArg());
  if (LLVM_UNLIKELY(!self)) {
    return runtime.raiseTypeError(
        "Non-Map object called on Map.prototype.size");
  }
  return HermesValue::encodeTrustedNumberValue(self->size());
}

CallResult<HermesValue> mapPrototypeValues(void *, Runtime &runtime) {
  NativeArgs args = runtime.getCurrentFrame().getNativeArgs();
  auto selfHandle = args.dyncastThis<JSMap>();
  if (LLVM_UNLIKELY(!selfHandle)) {
    return runtime.raiseTypeError(
        "Non-Map object called on Map.prototype.values");
  }
  struct : public Locals {
    PinnedValue<JSMapIterator> iterator;
  } lv;
  LocalsRAII lraii(runtime, &lv);
  lv.iterator.castAndSetHermesValue<JSMapIterator>(
      JSMapIterator::create(
          runtime, Handle<JSObject>::vmcast(&runtime.mapIteratorPrototype))
          .getHermesValue());
  lv.iterator->initializeIterator(runtime, selfHandle, IterationKind::Value);
  return lv.iterator.getHermesValue();
}

HermesValue createMapIteratorPrototype(Runtime &runtime) {
  struct : public Locals {
    PinnedValue<JSObject> parentHandle;
  } lv;
  LocalsRAII lraii(runtime, &lv);
  lv.parentHandle.castAndSetHermesValue<JSObject>(
      JSObject::create(
          runtime, Handle<JSObject>::vmcast(&runtime.iteratorPrototype))
          .getHermesValue());
  defineMethod(
      runtime,
      lv.parentHandle,
      Predefined::getSymbolID(Predefined::next),
      nullptr,
      mapIteratorPrototypeNext,
      0);

  auto dpf = DefinePropertyFlags::getDefaultNewPropertyFlags();
  dpf.writable = 0;
  dpf.enumerable = 0;
  defineProperty(
      runtime,
      lv.parentHandle,
      Predefined::getSymbolID(Predefined::SymbolToStringTag),
      runtime.getPredefinedStringHandle(Predefined::MapIterator),
      dpf);

  return lv.parentHandle.getHermesValue();
}

CallResult<HermesValue> mapIteratorPrototypeNext(void *, Runtime &runtime) {
  NativeArgs args = runtime.getCurrentFrame().getNativeArgs();
  auto selfHandle = args.dyncastThis<JSMapIterator>();
  if (LLVM_UNLIKELY(!selfHandle)) {
    return runtime.raiseTypeError(
        "Non-MapIterator object called on MapIterator.prototype.next");
  }

  auto cr = JSMapIterator::nextElement(selfHandle, runtime);
  if (LLVM_UNLIKELY(cr == ExecutionStatus::EXCEPTION)) {
    return ExecutionStatus::EXCEPTION;
  }
  return *cr;
}
} // namespace vm
} // namespace hermes

File: test/hermes/map.js
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

// RUN: %hermes -O %s | %FileCheck --match-full-lines %s
// RUN: %hermes -O -emit-binary -out %t.hbc %s && %hermes %t.hbc | %FileCheck --match-full-lines %s


print("Map prototype and constructor");
//CHECK-LABEL: Map prototype and constructor

print(Map.prototype);
//CHECK-NEXT: [object Map]
print(Map.length);
//CHECK-NEXT: 0

print(Object.getOwnPropertyNames(Map.prototype));
//CHECK-NEXT: clear,delete,entries,forEach,get,has,keys,set,size,values,constructor

print(new Map());
//CHECK-NEXT: [object Map]
var x = new Map(new Set([[1,2], [3,4]]));
print(x.get(1), x.get(3));
// CHECK-NEXT: 2 4

try {
  Map();
} catch (e) {
  print(e);
}
//CHECK-NEXT: TypeError: Constructor Map requires 'new'

try{ Map.prototype.clear(); } catch (e) { print(e); }
//CHECK-NEXT: TypeError: Non-Map object called on Map.prototype.clear

var m = new Map([]);
print(m.size);
// CHECK-NEXT: 0
var m = new Map([[1,2]]);
print(m.size, m.get(1));
// CHECK-NEXT: 1 2
var m = new Map([[1,2], [3,4]]);
print(m.size, m.get(1), m.get(3));
// CHECK-NEXT: 2 2 4
try { new Map([null]); } catch (e) { print('caught', e.name) }
// CHECK-NEXT: caught TypeError

var oldSet = Map.prototype.set;
Map.prototype.set = 123;
try { new Map([[1,2]]); } catch (e) { print('caught', e.name) }
// CHECK-NEXT: caught TypeError
var arr = [[1,2], [3,4]];
Map.prototype.set = function(k, v) {
  oldSet.call(this, k, v);
  arr.length = 1;
}
print(new Map(arr).size);
// CHECK-NEXT: 1
Map.prototype.set = function(k, v) {
  oldSet.call(this, k, v);
  // Changing the set function mid-construction shouldn't affect the call.
  Map.prototype.set = 10;
};
var m = new Map([[1,2], [3,4]]);
print(m.size, m.get(1), m.get(3));
// CHECK-NEXT: 2 2 4
Map.prototype.set = oldSet;

// Make sure we can insert internal object ID into a frozen object.
var fo = {};
Object.freeze(fo);
var fs = new Map();
fs.set(fo, fo);

function testSet(o1, o2, o3) {
  var s = new Map();
  print("testSet");
//CHECK-LABEL: testSet
  print(s.size);
//CHECK-NEXT: 0
  s.set(undefined, 0);
  s.set("abc", 1);
  s.set("def", 2);
  s.set("abc", 3);
  s.set("def", 4);
  s.set(1/-Infinity, 5);
  s.set(o1, 6);
  s.set(o2, 7);
  s.set(o3, 8);
  var o11 = o1, o22 = o2, o33 = o3;
  s.set(o11, 9);
  s.set(o22, 10);
  s.set(o33, 11);
  print(s.size);
  //CHECK-NEXT: 7
  return s;
}

function testHas(s, o1, o2, o3) {
  print("testHas");
//CHECK-LABEL: testHas
  print(s.has(0));
//CHECK-NEXT: true
  print(s.has(1));
//CHECK-NEXT: false
  print(s.has("abc"));
//CHECK-NEXT: true
  print(s.has("a"));
//CHECK-NEXT: false
  print(s.has("def"));
//CHECK-NEXT: true
  print(s.has());
//CHECK-NEXT: true
  print(s.has(null));
//CHECK-NEXT: false
  print(s.has(o1));
//CHECK-NEXT: true
  print(s.has(o2));
//CHECK-NEXT: true
  print(s.has(o3));
//CHECK-NEXT: true
  print(s.has({}));
//CHECK-NEXT: false
  var o4 = o2;
  print(s.has(o4));
//CHECK-NEXT: true
}

function testGet(s, o1, o2, o3) {
  print("testGet");
//CHECK-LABEL: testGet
  print(s.get(0));
//CHECK-NEXT: 5
  print(s.get(1));
//CHECK-NEXT: undefined
  print(s.get("abc"));
//CHECK-NEXT: 3
  print(s.get("a"));
//CHECK-NEXT: undefined
  print(s.get("def"));
//CHECK-NEXT: 4
  print(s.get());
//CHECK-NEXT: 0
  print(s.get(null));
//CHECK-NEXT: undefined
  print(s.get(o1));
//CHECK-NEXT: 9
  print(s.get(o2));
//CHECK-NEXT: 10
  print(s.get(o3));
//CHECK-NEXT: 11
  print(s.get({}));
//CHECK-NEXT: undefined
  var o4 = o2;
  print(s.get(o4));
//CHECK-NEXT: 10
}



function testDelete(s, o1, o2, o3) {
  print("testDelete");
//CHECK-LABEL: testDelete
  print(s.delete(0));
//CHECK-NEXT: true
  print(s.size);
//CHECK-NEXT: 6
  print(s.delete(1));
//CHECK-NEXT: false
  print(s.delete("abc"));
//CHECK-NEXT: true
  print(s.size);
//CHECK-NEXT: 5
  print(s.delete("a"));
//CHECK-NEXT: false
  print(s.delete("def"));
//CHECK-NEXT: true
  print(s.size);
//CHECK-NEXT: 4
  print(s.delete());
//CHECK-NEXT: true
  print(s.size);
//CHECK-NEXT: 3
  print(s.delete(null));
//CHECK-NEXT: false
  print(s.delete(o1));
//CHECK-NEXT: true
  print(s.size);
//CHECK-NEXT: 2
  var o4 = o2;
  print(s.delete(o4));
//CHECK-NEXT: true
  print(s.size);
//CHECK-NEXT: 1
  print(s.delete(o3));
//CHECK-NEXT: true
  print(s.size);
//CHECK-NEXT: 0
  print(s.delete({}));
//CHECK-NEXT: false

  print(s.has(0));
//CHECK-NEXT: false
  print(s.has("abc"));
//CHECK-NEXT: false
  print(s.has("def"));
//CHECK-NEXT: false
  print(s.has());
//CHECK-NEXT: false
  print(s.has(o1));
//CHECK-NEXT: false
  print(s.has(o2));
//CHECK-NEXT: false
  print(s.has(o3));
//CHECK-NEXT: false
}

function testClear(o1, o2, o3) {
  var s = testSet(o1, o2, o3);

  print("testClear");
//CHECK-LABEL: testClear
  print(s.size);
//CHECK-NEXT: 7
  s.clear();
  print(s.size);
//CHECK-NEXT: 0
}

function testIteration() {
  s = new Map();
  s.set(1, {});
  s.set({}, 1);
  s.set(undefined, null);
  s.set("abc", "def");

  print("testIteration");
//CHECK-LABEL: testIteration
  var keys = s.keys();
  print(keys.__proto__);
//CHECK-NEXT: [object Map Iterator]
  print(keys[Symbol.iterator]());
//CHECK-NEXT: [object Map Iterator]
  print(Object.getOwnPropertyNames(keys.__proto__));
//CHECK-NEXT: next
  while (true) {
    var e = keys.next();
    print(e.value);
    if (e.done) break;
  }
//CHECK-NEXT: 1
//CHECK-NEXT: [object Object]
//CHECK-NEXT: undefined
//CHECK-NEXT: abc
//CHECK-NEXT: undefined

  var values = s.values();
  while (true) {
    var e = values.next();
    print(e.value);
    if (e.done) break;
  }
//CHECK-NEXT: [object Object]
//CHECK-NEXT: 1
//CHECK-NEXT: null
//CHECK-NEXT: def
//CHECK-NEXT: undefined

  var entries = s.entries();
  while (true) {
    var e = entries.next();
    print(e.value);
    if (e.done) break;
  }
//CHECK-NEXT: 1,[object Object]
//CHECK-NEXT: [object Object],1
//CHECK-NEXT: ,
//CHECK-NEXT: abc,def
//CHECK-NEXT: undefined
}

function testForEach() {
  function callbackFn(value, key, s) {
    print(key, value, s.size);
  }
  s = new Map();
  s.set(1, []);
  s.set({}, -0);
  s.set(undefined, undefined);
  s.set("abc", "");

  print("testForEach", Map.prototype.forEach.length);
//CHECK-LABEL: testForEach 1
  s.forEach(callbackFn);
//CHECK-NEXT: 1  4
//CHECK-NEXT: [object Object] 0 4
//CHECK-NEXT: undefined undefined 4
//CHECK-NEXT: abc  4
}

function testZero() {
  print("testZero");
//CHECK-LABEL: testZero

  // -0 and +0 are normalized to +0
  var m = new Map();
  m.set(-0, 42);
  print(m.has(-0));
  print(m.has(+0));
  print(m.get(-0));
  print(m.get(+0));
  print(Object.is(-0, [...m.keys()][0]));
  print(Object.is(+0, [...m.keys()][0]));
  print(m.delete(+0))
  print(m.size === 0)
//CHECK-NEXT: true
//CHECK-NEXT: true
//CHECK-NEXT: 42
//CHECK-NEXT: 42
//CHECK-NEXT: false
//CHECK-NEXT: true
//CHECK-NEXT: true
//CHECK-NEXT: true
  m.set(+0, 43);
  print(m.has(-0));
  print(m.has(+0));
  print(m.get(-0));
  print(m.get(+0));
  print(Object.is(-0, [...m.keys()][0]));
  print(Object.is(+0, [...m.keys()][0]));
  print(m.delete(-0))
  print(m.size === 0)
//CHECK-NEXT: true
//CHECK-NEXT: true
//CHECK-NEXT: 43
//CHECK-NEXT: 43
//CHECK-NEXT: false
//CHECK-NEXT: true
//CHECK-NEXT: true
//CHECK-NEXT: true

  // Only key should be normalized.
  var m = new Map();
  m.set(-0, -0);
  var [k, v] = m.entries().next().value;
  print(Object.is(-0, k));
  print(Object.is(-0, v));
//CHECK-NEXT: false
//CHECK-NEXT: true
}

var o1 = {};
var o2 = {a: 1};
var o3 = {a: 1, b: 2};
var s = testSet(o1, o2, o3);
testHas(s, o1, o2, o3);
testGet(s, o1, o2, o3);
testDelete(s, o1, o2, o3);
testClear(o1, o2, o3);
testIteration();
testForEach();
testZero();
